presto-orc/src/main/java/com/facebook/presto/orc/reader/ByteSelectiveStreamReader.java

```diff
@@ -264,7 +277,20 @@ else if (nullsAllowed) {
         return positions[positionCount - 1] + 1;
     }
 
-    private int readNoFilter(int[] positions, int positionCount)
+    private int readNoFilterNoNull(int[] positions, int positionCount)
+            throws IOException
+    {
+        // filter == null implies outputRequired == true
+        int totalPositionCount = positions[positionCount - 1] + 1;
+        nonNullValueBuffer = ensureCapacity(nonNullValueBuffer, totalPositionCount);
+        dataStream.next(nonNullValueBuffer, totalPositionCount);
```

Comment:
@yingsu00 

- This new interface is confusing; the signature suggests that it reads next `totalPositionCount` items, however, the implementation is different. It reads bytes up to `totalPositionCount`-th byte.
- Allocating buffer for `totalPositionCount` bytes when we need just `positionCount` is wasteful. Copying bytes with System.arraycopy then copying again with ReaderUtils.packByteValues also seems inefficient. How about, replacing the new interface with the one that takes positions? This way it won't be necessary to allocate additional `nonNullValueBuffer` array and the data will be copied only once. It may also make sense to throw in an optimization of using System.arraycopy if all positions are needed, e.g. `positions[positionCount - 1] - positions[0] = positionCount - 1`.

Reply:
@mbasmanova thanks for reviewing. For your second question, I actually did try to push the positions[] array to the dataStream and do skipping in the InputStream level but the gain was not obvious. But in this implementation in this PR, although we did allocate more and copy more, but the the gain of doing batch processing and System.arraycopy is fast enough to offset the disadvantages of doing one more copy and larger allocation. I did experiments which showed that up to 100x reduction rate by positions[] array, this implementation wins. How about I add a reduction rate check such that when the rate is below 10x that we use this implementation, while >10x we fall back to the previous solution?
cc @oerling

Reply:
@yingsu00 I'm not saying that the modifications will result in better performance. That may not be the case like you pointed out. However, the code will be cleaner and shorter and won't be as confusing to the reader as it is now.

Reply:
@mbasmanova The `dataStream.next(nonNullValueBuffer, totalPositionCount);` does read totalPositionCount bytes to nonNullValueBuffer. Am I misunderstanding something? 

How about the following code? 

```
private int readNoFilterNoNull(int[] positions, int positionCount)
            throws IOException
    {
        // filter == null implies outputRequired == true
        int totalPositionCount = positions[positionCount - 1] + 1;
        if (totalPositionCount == positionCount) {
            dataStream.next(values, positionCount);
            return positionCount;
        }

        int streamPosition = 0;
        for (int i = 0; i < positionCount; i++) {
            int position = positions[i];
            if (position > streamPosition) {
                skip(position - streamPosition);
                streamPosition = position;
            }

            values[i] = dataStream.next();

            streamPosition++;
        }

        outputPositionCount = positionCount;
        return streamPosition;
    }

```

Reply:
@yingsu00 

> The dataStream.next(nonNullValueBuffer, totalPositionCount); does read totalPositionCount bytes to nonNullValueBuffer.

You are correct. I misread the code in that method. Thanks for clarifying.

What's the gist of this change? Is it to add bulk read method to ByteInputStream? E.g. instead of reading one byte a  a time, copy a whole range of bytes at once? This change would optimize the case of reading contiguous chunks of rows of TINYINTs with no filters and no nulls, right? Is this an important use case to optimize?

If my understanding is accurate, then I suggest to make a single commit `Optimize reading of contiguous TINYINTs with no filter and no nulls` that introduces and uses bulk `next` method when the conditions are met (no nulls, no filter, no skipping).

Reply:
@mbasmanova Yes you're correct. Adding bulk read to the InputStream then unpack the read bytes allows tight loop to be used and improves the performance. However it uses additional memory as you pointed out, so we'll just limit the use case to just `totalPositionCount == positionCount`. 
About the benefit, remember there's 38% of total warehouse scan CPU are not filtering anything (filter rate = 1). https://fb.quip.com/PdX6As714xYv 
I will make it one commit. Thank you for the suggestions!

Reply:
@yingsu00 

> About the benefit, remember there's 38% of total warehouse scan CPU are not filtering anything (filter rate = 1)

But this change only applies to TINYINT columns, right? I didn't think these occur often.

Reply:
This is just the first type. The same idea applies to other types and I will make separate PRs for other types.